src/shell/js/composeui-fdc3/src/ComposeUIDesktopAgent.ts

```diff
@@ -0,0 +1,312 @@
+        throw new Error("Not implemented");
+    }
+
+    //TODO
+    public getSystemChannels(): Promise<Channel[]> {
+        throw new Error("Not implemented");
+    }
+
+    //TODO: Revisit for private channels
+    public joinChannel(channelId: string): Promise<void> {
+        return new Promise<void>((resolve, reject) => {
+            if (this.currentChannel) {
+                reject(new Error(ChannelError.CreationFailed));
+                return;
+            }
+
+            let channel = this.findChannel(channelId, "user");
+            if (channel) {
+                this.currentChannel = channel;
+                resolve();
+                return;
+            }
+
+            channel = this.findChannel(channelId, "app");
+            if (channel) {
+                this.currentChannel = channel;
+                resolve();
+                return;
+            }
+
+            channel = this.findChannel(channelId, "private");
+            if (channel) {
+                this.currentChannel = channel;
+                resolve();
+                return;
+            }
+
+            if (!channel) {
+                reject(new Error(`No channel is found with id: ${channelId}`));
+                return;
+            }
+        });
+    }
+
+    private findChannel(channelId: string, channelType: ChannelType): ComposeUIChannel | undefined {
+        let channel: ComposeUIChannel | undefined;
+        const predicate = (channel: Channel) => channel.id == channelId;
+
+        switch (channelType) {
+            case "app":
+                channel = this.appChannels.find(predicate);
+                break;
+            case "private":
+                channel = this.privateChannels.find(predicate);
+                break;
+            case "user":
+                channel = this.userChannels.find(predicate);
+                break;
+        }
+
+        return channel;
+    }
+
+    private addChannel(channel: ComposeUIChannel): void {
+        if (channel == null) return;
+        switch (channel.type) {
+            case "app":
+                this.appChannels.push(channel);
+                break;
+            case "user":
+                this.userChannels.push(channel);
+                break;
+            case "private":
+                this.privateChannels.push(channel);
+                break;
+        }
+    }
+
+    private async invokeChannelCreationMessage(topic: string, channelId: string, channelType: ChannelType): Promise<void> {
+        const message = JSON.stringify(new Fdc3FindChannelRequest(channelId, channelType));
+        await this.messageRouterClient.invoke(topic, message)
+                    .then((response) => {
+                        if(response) {
+                            const message = JSON.parse(response) as TopicMessage;
+                            if(message.payload) {
+                                const fdc3Message = JSON.parse(message.payload) as Fdc3ErrorResponse;
```

Comment:
That won't work, you won't get an object that's actually an `Fc3ErrorResponse`, you can't deserialize a class like that. The `as` operator in TypeScript just tells the compiler to treat the value *as if* it was of the specified type. The response should be declared as an interface instead.